### PR TITLE
perf(atoms): Introduce `AtomStoreCell`

### PR DESCRIPTION
**Pull request by dsherret**
Introduces an `AtomStoreCell` type which internally uses an `UnsafeCell`, but enforces a safe contract. This allows skipping the borrow check of `RefCell`.

`> cargo "bench" "--package" "swc_ecma_parser" "--bench" "parser" "--" "three"`

Main: ~37.8ms
This PR: ~37.4ms

### CODE REVIEW

File: crates/swc_ecma_parser/src/lexer/util.rs [hunk 1]
```diff
@@ -22,6 +23,19 @@ use crate::{
...
 };
 
+#[derive(Default)]
+pub(crate) struct AtomStoreCell(std::cell::UnsafeCell<AtomStore>);
+
+impl AtomStoreCell {
+    #[inline]
+    pub fn atom<'a>(&self, s: impl Into<Cow<'a, str>>) -> Atom {
+        // SAFETY: We can skip the borrow check of RefCell because
+        // this API enforces a safe contract. It is slightly faster
```
[the remark this answers is not in the document]
Comment (dsherret) [reply crates/swc_ecma_parser/src/lexer/util.rs:33]: Is it possible to use the current API of AtomStoreCell in an unsafe way?

[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc_ecma_parser/src/lexer/util.rs:33]: Ah you are right. The mutable borrow ends in the function.

Can you add this to the comment?

[the remark this answers is not in the document]
Comment (dsherret) [reply crates/swc_ecma_parser/src/lexer/util.rs:33]: Sure, I will do it in the morning here.

**Comment by kdy1** [comment]
Maybe we can move this to `swc_atoms`

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_atoms

**Pull request merged**